File: packages/shared-ui/src/interface/cells.ts
import type { AirDamage, Battle } from "./battle";

export interface Cells {
  maparea_id: number;
  mapinfo_no: number;
  bosscell_no: number;
  bosscomp: number;
  cells: { [key: number]: Cell };
  cell_index: number[];
  event_map?: Eventmap;
  cell_data: CellData[];
  // timestamp: number,
  battles: { [key: number]: Battle };
}

export interface Cell {
  rashin_id: number;
  no: number;
  color_no: number;
  event_id: number;
  event_kind: number;
  next: number;
  e_deck_info?: EDeckInfo[];
  limit_state: number;
  m1?: number;
  destruction_battle?: DestructionBattle;
  happening?: Happening;
}

export interface CellData {
  id: number;
  no: number;
  color_no: number;
  passed: number;
  distance?: number;
}

export interface Eventmap {
  max_maphp: number;
  now_maphp: number;
  dmg: number;
}

export interface Happening {
  count: number;
  mst_id: number;
  dentan: number;
}

export interface EDeckInfo {
  kind: number;
  ship_ids: number[];
}

export interface DestructionBattle {
  formation: number[];
  ship_ke: number[];
  ship_lv: number[];
  e_nowhps: number[];
  e_maxhps: number[];
  e_slot: number[][];
  f_nowhps: number[];
  f_maxhps: number[];
  air_base_attack: ApiAirBaseAttack;
  lost_kind: number;
  f_total_damages: number[] | null;
  e_total_damages: number[] | null;
}

export interface ApiAirBaseAttack {
  air_superiority: number | null;
  stage_flag: number[];
  squadron_plane: (number | null)[] | null;
  f_damage: AirDamage;
  e_damage: AirDamage;
  map_squadron_plane: { [key: string]: number[] } | null;
}

export const default_cells: Cells = {
  maparea_id: 0,
  mapinfo_no: 0,
  bosscell_no: 0,
  bosscomp: 0,
  cells: {},
  cell_index: [],
  cell_data: [],
  battles: {},
  // timestamp: 0,
};

File: packages/kc_api/crates/kc-api-interface/src/cells.rs
use once_cell::sync::Lazy;
use std::collections::HashMap;
use std::sync::Mutex;

use super::battle::{AirDamage, Battle};

use serde::{Deserialize, Serialize};
use ts_rs::TS;

pub static KCS_CELLS_INDEX: Lazy<Mutex<Vec<i64>>> = Lazy::new(|| Mutex::new(Vec::new()));
pub static KCS_CELLS: Lazy<Mutex<Cells>> = Lazy::new(|| {
    Mutex::new(Cells {
        maparea_id: 0,
        mapinfo_no: 0,
        bosscell_no: 0,
        bosscomp: 0,
        cells: HashMap::new(),
        cell_index: Vec::new(),
        event_map: None,
        cell_data: Vec::new(),
        battles: HashMap::new(),
    })
});

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "cells.ts")]
pub struct Cells {
    pub maparea_id: i64,
    pub mapinfo_no: i64,
    pub bosscell_no: i64,
    pub bosscomp: i64,
    pub cells: HashMap<i64, Cell>,
    pub cell_index: Vec<i64>,
    pub event_map: Option<Eventmap>,
    pub cell_data: Vec<CellData>,
    pub battles: HashMap<i64, Battle>,
}

impl Cells {
    pub fn load() -> Self {
        let cells = KCS_CELLS.lock().unwrap();
        cells.clone()
    }

    pub fn restore(&self) {
        let mut cells = KCS_CELLS.lock().unwrap();
        *cells = self.clone();
    }

    pub fn reset() {
        let mut cells = KCS_CELLS.lock().unwrap();
        cells.cells.clear();
        cells.cell_index.clear();
        cells.event_map = None;
        cells.cell_data.clear();
        cells.battles.clear();
        cells.maparea_id = 0;
        cells.mapinfo_no = 0;
        cells.bosscell_no = 0;
        cells.bosscomp = 0;
    }

    pub fn reset_flag() -> bool {
        let cells = KCS_CELLS.lock().unwrap();
        cells.maparea_id == 0
            && cells.mapinfo_no == 0
            && cells.bosscell_no == 0
            && cells.bosscomp == 0
            && cells.cells.is_empty()
            && cells.cell_index.is_empty()
            && cells.event_map.is_none()
            && cells.cell_data.is_empty()
            && cells.battles.is_empty()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "cells.ts")]
pub struct Cell {
    pub timestamp: Option<i64>,
    pub rashin_id: i64,
    pub no: i64,
    pub color_no: i64,
    pub event_id: i64,
    pub event_kind: i64,
    pub next: i64,
    // pub event_id: i64,
    // pub event_kind: i64,
    // pub airsearch: Airsearch,
    // pub comment_kind: Option<i64>,
    // pub production_kind: Option<i64>,
    pub e_deck_info: Option<Vec<EDeckInfo>>,
    pub limit_state: i64,
    // pub ration_flag: Option<i64>,
    // pub select_route: Option<SelectRoute>,
    // pub itemget: Option<Vec<Itemget>>,
    pub m1: Option<i64>,
    pub destruction_battle: Option<DestructionBattle>,
    pub happening: Option<Happening>,
}

impl Cell {
    pub fn add_or(&self) {
        let mut cells = KCS_CELLS.lock().unwrap();
        cells.cells.insert(self.no, self.clone());
        cells.cell_index.push(self.no);
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "cells.ts")]
pub struct CellData {
    pub id: i64,
    pub no: i64,
    pub color_no: i64,
    pub passed: i64,
    pub distance: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "cells.ts")]
pub struct Eventmap {
    pub max_maphp: i64,
    pub now_maphp: i64,
    pub dmg: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "cells.ts")]
pub struct Happening {
    // type: i64,
    pub count: i64,
    // usemst: i64,
    pub mst_id: i64,
    // icon_id: i64,
    pub dentan: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "cells.ts")]
pub struct EDeckInfo {
    pub kind: i64,
    pub ship_ids: Vec<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "cells.ts")]
pub struct DestructionBattle {
    pub formation: Vec<i64>,
    pub ship_ke: Vec<i64>,
    pub ship_lv: Vec<i64>,
    pub e_nowhps: Vec<i64>,
    pub e_maxhps: Vec<i64>,
    pub e_slot: Vec<Vec<i64>>,
    pub f_nowhps: Vec<i64>,
    pub f_maxhps: Vec<i64>,
    // Need to implement
    pub air_base_attack: AirBaseAttack,
    pub lost_kind: i64,
    pub f_total_damages: Option<Vec<i64>>,
    pub e_total_damages: Option<Vec<i64>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, TS)]
#[ts(export, export_to = "cells.ts")]
pub struct AirBaseAttack {
    pub air_superiority: Option<i64>,
    pub plane_from: Vec<Option<Vec<i64>>>,
    pub f_damage: AirDamage,
    pub e_damage: AirDamage,
    pub stage_flag: Vec<i64>,
    pub map_squadron_plane: Option<HashMap<String, Vec<i64>>>,
}
